quickwit/quickwit-doc-mapper/src/default_doc_mapper/mapping_tree.rs

```diff
@@ -214,34 +183,53 @@ impl MappingLeaf {
     }
 }
 
-fn json_type_from_json_value(json_value: &JsonValue) -> JsonType {
-    match json_value {
-        JsonValue::Null => JsonType::Null,
-        JsonValue::Bool(_) => JsonType::Bool,
-        JsonValue::Array(_) => JsonType::Array,
-        JsonValue::Number(_) => JsonType::Number,
-        JsonValue::String(_) => JsonType::String,
-        JsonValue::Object(_) => JsonType::Object,
-    }
-}
-
 fn extract_json_val(
-    json_type: JsonType,
-    named_doc: &mut BTreeMap<String, Vec<JsonValue>>,
+    leaf_type: &LeafType,
+    named_doc: &mut BTreeMap<String, Vec<Value>>,
     field_path: &[&str],
     cardinality: Cardinality,
 ) -> Option<JsonValue> {
     let full_path = field_path.join(".");
     let vals = named_doc.remove(&full_path)?;
     let mut vals_with_correct_type_it = vals
         .into_iter()
-        .filter(|json_val| json_type_from_json_value(json_val) == json_type);
+        .filter_map(|value| value_to_json(value, leaf_type));
     match cardinality {
         Cardinality::SingleValue => vals_with_correct_type_it.next(),
         Cardinality::MultiValues => Some(JsonValue::Array(vals_with_correct_type_it.collect())),
     }
 }
 
+fn value_to_json(value: Value, leaf_type: &LeafType) -> Option<JsonValue> {
+    // Make sure the requested type and the value type are consistent.
+    let matched_value = match (&value, leaf_type) {
+        (Value::Str(_), LeafType::Text(_)) => value,
+        (Value::I64(_), LeafType::I64(_)) => value,
+        (Value::U64(_), LeafType::U64(_)) => value,
+        (Value::F64(_), LeafType::F64(_)) => value,
+        (Value::Bool(_), LeafType::Bool(_)) => value,
+        (Value::IpAddr(_), LeafType::IpAddr(_)) => value,
+        (Value::Date(_), LeafType::DateTime(_)) => value,
+        (Value::Bytes(_), LeafType::Bytes(_)) => value,
+        (Value::JsonObject(_), LeafType::Json(_)) => value,
+        _ => return None,
+    };
+
+    // Apply user requested formatting.
+    if let (Value::Date(date_time), LeafType::DateTime(date_time_options)) =
+        (&matched_value, leaf_type)
+    {
+        let json_value = date_time_options
+            .format_to_json(*date_time)
+            .expect("Invalid datetime is not allowed.");
+        return Some(json_value);
+    }
+
+    let json_value =
```

Comment:
Question: for bytes, we need to encode into base64 as we want to return that to the user. Where is the encoding part done?

Reply:
TantivyValue::Bytes already serializes it to base64